Cython/Compiler/ExprNodes.py

```diff
@@ -12897,15 +12897,25 @@ def coerce_to_integer(self, env):
         if not self.false_val.type.is_int:
             self.false_val = self.false_val.coerce_to_integer(env)
         self.result_ctype = None
-        return self.analyse_result_type(env)
+        out = self.analyse_result_type(env)
+        if not out.type.is_int:
+            # fall back to ordinary coercion since we haven't ended as the correct type
+            assert isinstance(out, CondExprNode)
+            out = super(CondExprNode, out).coerce_to_integer(env)
```

Comment:
Instead of an assertion here, I guess we could also check if `out is not self` and then call `out.coerce_to_integer()`. (Same below.)

Besides, I checked where we use `coerce_to_integer()` and it's mostly cases where we'd really want `Py_ssize_t` as type rather than `long`, which we currently let it coerce to. We should probably change it to `Py_ssize_t`.

Reply:
I've replaces the assertion here.

Not sure I quite understand your point about `Py_ssize_t` rather than long? Are you proposing a general change of `coerce_to_integer` to `coerce_to_py_ssize_t` everywhere? I can't see anywhere specific to `CondExprNode` that uses `long`.

Reply:
Sorry, that was a general remark, not something to fix in this PR. On Windows, `long` is 32 bits, so using `coerce_to_integer()` for index values (as we do) is actually dangerous. It should coerce to `Py_ssize_t` instead. And we probably also shouldn't use it to coerce to specific ints but use `coerce_to()` in that case. It feels like a special case that is not special enough to merit its own method. But as you suggest, a `coerce_to_index()` might be helpful since we use that intention in several places.

I created https://github.com/cython/cython/issues/5743 for this.